# Naïve Bayes

### 1. Import Library dan Load Dataset Iris

In [1]:
import numpy as np
import pandas as pd
from sklearn.datasets import load_iris

# Load dataset Iris dari scikit-learn
iris = load_iris()

numpy digunakan untuk pengolahan data numerik.
pandas digunakan untuk membuat dan mengolah DataFrame.
load_iris digunakan untuk mengambil dataset Iris dari scikit-learn.

Kode load dataset Iris mengambil dataset Iris yang sudah tersedia di scikit-learn. Dataset ini berisi 150 data bunga Iris dengan empat fitur.

### 2. Mengubah Dataset Menjadi DataFrame dan Menentukan Fitur X dan Target y

In [2]:
# Konversi data menjadi DataFrame Pandas
df = pd.DataFrame(iris.data, columns=iris.feature_names)
df["Species"] = iris.target_names[iris.target]

# Definisikan X (fitur) dan y (target/label)
X = df.drop(columns=["Species"])
y = df["Species"]

print(X)

     sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
0                  5.1               3.5                1.4               0.2
1                  4.9               3.0                1.4               0.2
2                  4.7               3.2                1.3               0.2
3                  4.6               3.1                1.5               0.2
4                  5.0               3.6                1.4               0.2
..                 ...               ...                ...               ...
145                6.7               3.0                5.2               2.3
146                6.3               2.5                5.0               1.9
147                6.5               3.0                5.2               2.0
148                6.2               3.4                5.4               2.3
149                5.9               3.0                5.1               1.8

[150 rows x 4 columns]


Data fitur dari dataset Iris dimasukkan ke dalam DataFrame. Kemudian ditambahkan kolom Species yang berisi jenis bunga Iris.

Di sini data dipisahkan menjadi dua bagian:
X (data fitur yang digunakan untuk melakukan prediksi.)
y (target atau hasil yang ingin diprediksi, yaitu Species.)

Jadi, model akan mempelajari hubungan antara ukuran sepal dan petal dengan jenis bunga Iris.

### 3. Menampilkan Target atau Label Species pada Dataset Iris

In [3]:
y

0         setosa
1         setosa
2         setosa
3         setosa
4         setosa
         ...    
145    virginica
146    virginica
147    virginica
148    virginica
149    virginica
Name: Species, Length: 150, dtype: str

y adalah variabel target yang nantinya akan diprediksi oleh model Naïve Bayes, sedangkan data pada X berisi fitur-fitur yang digunakan untuk melakukan prediksi. Di modul, y memang didefinisikan sebagai kolom Species.

### 4. Mengubah Label Species Menjadi Angka

In [4]:
from sklearn.preprocessing import LabelEncoder
lab = LabelEncoder()
y = lab.fit_transform(y)
y

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2])

LabelEncoder digunakan untuk mengubah label berbentuk teks menjadi angka agar dapat diproses oleh model.

### 5. Membagi Data Training dan Testing

In [5]:
from sklearn.model_selection import train_test_split
X_train, X_test = train_test_split(X, test_size=0.2, random_state=0)
y_train, y_test = train_test_split(y, test_size=0.2, random_state=0)


Dataset dibagi menjadi dua:

80% data training, digunakan untuk melatih model.
20% data testing, digunakan untuk menguji kemampuan model.

random_state=0 digunakan supaya pembagian datanya tetap sama ketika kode dijalankan kembali.

### 6. Membuat dan Melatih Model

In [6]:
from sklearn.naive_bayes import GaussianNB
model = GaussianNB()

NB_model = model.fit(X_train, y_train)

GaussianNB() digunakan untuk membuat model Gaussian Naïve Bayes.

### 7. Melakukan Prediksi Data Testing

In [7]:
result = NB_model.predict(X_test)
result

array([2, 1, 0, 2, 0, 2, 0, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 0, 2, 1,
       0, 0, 2, 0, 0, 1, 1, 0])

Kode tersebut meminta model untuk memprediksi species berdasarkan data testing.

### 8. Menghitung Accuracy

In [8]:
from sklearn.metrics import accuracy_score
print('Accuracy of data testing: ', accuracy_score(y_test, result)*100)

Accuracy of data testing:  96.66666666666667


Accuracy digunakan untuk mengetahui berapa banyak prediksi model yang benar dibandingkan dengan seluruh data testing.
Pada hasil modul, accuracy yang diperoleh adalah 96.67%.
Artinya, dari data testing yang digunakan, sebagian besar species berhasil diprediksi dengan benar.

### 9. Prediksi Data Baru

In [9]:
#prediksi inputan data baru
#input_data = pd.DataFrame(np.array([[5.1, 3.5, 1.4, 0.2]]), columns=iris.feature_names)
input_data = [[5.1, 3.5, 1.4, 0.2]]
result = NB_model.predict(input_data)
prediction_label = lab.inverse_transform(result)
print('Hasil prediksi untuk input baru adalah:', prediction_label)

Hasil prediksi untuk input baru adalah: ['setosa']


c:\Users\LENOVO\AppData\Local\Programs\Python\Python313\Lib\site-packages\sklearn\utils\validation.py:2827: UserWarning: X does not have valid feature names, but GaussianNB was fitted with feature names
  warnings.warn(


Pada bagian ini dilakukan prediksi untuk data input baru menggunakan model Gaussian Naïve Bayes yang sudah dilatih sebelumnya. Data yang digunakan adalah [5.1, 3.5, 1.4, 0.2], yang terdiri dari nilai sepal length, sepal width, petal length, dan petal width. Data langsung dimasukkan dalam bentuk list, kemudian NB_model.predict() digunakan untuk menentukan species berdasarkan data tersebut. Selanjutnya, hasil prediksi yang masih berupa angka diubah menjadi nama species menggunakan lab.inverse_transform().

In [10]:
input_data = pd.DataFrame(np.array([[5.1, 3.5, 1.4, 0.2]]), columns=iris.feature_names)
result = NB_model.predict(input_data)
prediction_label = lab.inverse_transform(result)
print('Hasil prediksi untuk input baru adalah:', prediction_label)


Hasil prediksi untuk input baru adalah: ['setosa']


Data yang sama dibuat menjadi DataFrame dan diberikan nama kolom sesuai dengan fitur pada dataset Iris. Cara ini membuat format data input sama dengan data yang digunakan saat model dilatih dan menghindari warning mengenai nama fitur. Kedua cara tersebut menggunakan data yang sama sehingga menghasilkan prediksi species yang sama, yaitu setosa. Modul juga menunjukkan bahwa penggunaan DataFrame dengan nama kolom merupakan cara untuk menjaga struktur data tetap konsisten dengan model yang telah dilatih.

## Tugas

#### Prediksi data baru masuk ke species apa: [7.3, 3.8, 4.6, 2.3]

In [11]:
# Data baru yang akan diprediksi
input_data = pd.DataFrame(
    [[7.3, 3.8, 4.6, 2.3]],
    columns=iris.feature_names
)

# Melakukan prediksi
result = NB_model.predict(input_data)

# Mengubah hasil angka menjadi nama species
prediction_label = lab.inverse_transform(result)

print("Hasil prediksi untuk input baru adalah:", prediction_label)

Hasil prediksi untuk input baru adalah: ['virginica']


Pada tahap ini, dilakukan prediksi untuk data bunga Iris yang baru menggunakan model Gaussian Naïve Bayes yang sudah dilatih sebelumnya. Data yang digunakan adalah [7.3, 3.8, 4.6, 2.3], yang terdiri dari nilai panjang dan lebar sepal serta panjang dan lebar petal. Data tersebut dimasukkan ke dalam DataFrame agar formatnya sesuai dengan data yang digunakan saat proses pelatihan. Setelah itu, model melakukan prediksi menggunakan fungsi predict(). Hasil prediksi yang masih berupa angka kemudian diubah kembali menjadi nama species menggunakan inverse_transform(). Dari hasil tersebut, data baru [7.3, 3.8, 4.6, 2.3] diprediksi termasuk ke dalam species virginica.